### ISSUE TITLE
Pilot asks for 2-factor code in CI mode (TravisCI) and ignores FASTLANE_SESSION and FASTLANE_APPLE_APPLICATION_SPECIFIC_PASSWORD

### ISSUE THREAD

**Issue by Badmuts** [fastlane/fastlane#8298]
### New Issue Checklist

- [x] Updated fastlane to the latest version
- [x] I have read the [Contribution Guidelines](https://github.com/fastlane/fastlane/blob/master/CONTRIBUTING.md)

### Issue Description
Currently I am experiencing a problem with pilot and deliver running on TravisCI. I have a cordova project which creates build artifacts for Android and iOS. Fastlane uploads these artifacts via supply, pilot and deliver. The build process runs on TravisCI.

The problem I am experiencing is with pilot and deliver when fastlane is asking for a 2 factor authentication code. Off course I cannot provide this in my CI step. So I followed the [spaceship guide](https://github.com/fastlane/fastlane/tree/master/spaceship#support-for-ci-machines), created a `FASTLANE_SESSION` and `FASTLANE_APPLE_APPLICATION_SPECIFIC_PASSWORD` and provided these environment variables in my `.travis.yml` file.

When I run `bundle exec fastlane ios beta` locally the deployment runs without problems and does not ask for passwords or 2 factor auth codes. When TravisCI triggers a build Travis is prompted for a 6 digit code:

```
Login to iTunes Connect (me@email.com)
Two Factor Authentication for account 'me@email.com' is enabled
If you're running this in a non-interactive session (e.g. server or CI)
check out https://github.com/fastlane/fastlane/tree/master/spaceship#2-step-verification
Please enter the 6 digit code: 
```

Travis will timeout after this.

The iTunes Connect account I am using is part of multiple organisations. 

Removing `FASTLANE_PASSWORD` results in:

```
Missing password for user me@email.com, and running in non-interactive shell
```

First I thought this might be related to #8218 but I think these problems are not related because my match step is completed successful. 

I hope I provided enough information 😄 

##### Complete output when running fastlane, including the stack trace and command used
> You can use: `--capture_output` as the last commandline argument to get that collected for you

```
bundle exec fastlane ios beta
[12:56:09]: -------------------------------------------------
[12:56:09]: --- Step: Verifying required fastlane version ---
[12:56:09]: -------------------------------------------------
[12:56:09]: Your fastlane version 2.17.1 matches the minimum requirement of 2.9.0  ✅
[12:56:09]: ------------------------------
[12:56:09]: --- Step: default_platform ---
[12:56:09]: ------------------------------
[12:56:09]: Driving the lane 'ios beta' 🚀
[12:56:09]: -------------------
[12:56:09]: --- Step: pilot ---
[12:56:09]: -------------------
[12:56:10]: Login to iTunes Connect (me@email.com)
Two Factor Authentication for account 'me@email.com' is enabled
If you're running this in a non-interactive session (e.g. server or CI)
check out https://github.com/fastlane/fastlane/tree/master/spaceship#2-step-verification
Please enter the 6 digit code: 
```

### Environment

Please run `fastlane env` and copy the output below. This will help us help you :+1:
If you used `--capture_output` option please remove this block - as it is already included there.

<details><summary>✅ fastlane environment ✅</summary>

### Stack

| Key                 | Value                                       |
| ------------------- | ------------------------------------------- |
| OS                  | 10.11.6                                     |
| Ruby                | 2.3.3                                       |
| Bundler?            | true                                        |
| Git                 | git version 2.10.1 (Apple Git-78)           |
| Installation Source | ~/.rvm/gems/ruby-2.3.3/bin/fastlane         |
| Host                | Mac OS X 10.11.6 (15G1212)                  |
| Ruby Lib Dir        | ~/.rvm/rubies/ruby-2.3.3/lib                |
| OpenSSL Version     | OpenSSL 1.0.2k  26 Jan 2017                 |
| Is contained        | false                                       |
| Is homebrew         | false                                       |
| Xcode Path          | /Applications/Xcode.app/Contents/Developer/ |
| Xcode Version       | 8.2.1                                       |

### System Locale

| Variable | Value       |   |
| -------- | ----------- | - |
| LANG     | en_US.UTF-8 | ✅ |
| LC_ALL   |             |   |
| LANGUAGE |             |   |

### fastlane files:

<details><summary>`./fastlane/Fastfile`</summary>

```ruby
# Customise this file, documentation can be found here:
# https://github.com/fastlane/fastlane/tree/master/fastlane/docs
# All available actions: https://docs.fastlane.tools/actions
# can also be listed using the `fastlane actions` command

# Change the syntax highlighting to Ruby
# All lines starting with a # are ignored when running `fastlane`

# If you want to automatically update fastlane if a new version is available:
# update_fastlane

# This is the minimum version number required.
# Update this, if you use features of a newer version
fastlane_version "2.9.0"

default_platform :ios

desc "Retrieve certificates"
lane :certificates do
    if is_ci
        create_keychain(
            name: ENV["MATCH_KEYCHAIN_NAME"],
            password: ENV["MATCH_PASSWORD"],
            default_keychain: true,
            unlock: true,
            timeout: 3600,
            add_to_search_list: true
        )

        match(
            type: "development",
            keychain_name: ENV["MATCH_KEYCHAIN_NAME"],
            keychain_password: ENV["MATCH_PASSWORD"],
            readonly: true
        )

        match(
            type: "appstore",
            keychain_name: ENV["MATCH_KEYCHAIN_NAME"],
            keychain_password: ENV["MATCH_PASSWORD"],
            readonly: true
        )
    else
        match(type: "development", readonly: true)
        match(type: "appstore", readonly: true)
    end
end

platform :ios do
    before_all do
        ENV["SLACK_URL"] = "******"
    end

    desc "Submit a new Beta Build to Apple TestFlight"
    desc "This will also make sure the profile is up to date"
    lane :beta do
        pilot(
            ipa: "platforms/ios/build/device/app.ipa",
            skip_waiting_for_build_processing: true
        )
    end

    desc "Deploy a new version to the App Store"
    lane :release do
        deliver(
            ipa: "platforms/ios/build/device/app.ipa",
            force: true
        )
    end

    after_all do |lane|
        unless is_ci
            # This block is called, only if the executed lane was successful
            slack()
        end
    end

    error do |lane, exception|
        if is_ci
            slack(message: exception.message, success: false)
        end
    end
end

platform :android do
    before_all do
        ENV["SLACK_URL"] = "*****"
    end

    desc "Submit a new Beta Build to Google Play Store"
    desc "This will also make sure the profile is up to date"
    lane :beta do
        supply(
            apk: "platforms/android/build/outputs/apk/android-release.apk",
            package_name: "*****",
            json_key: "key.json",
            track: "beta"
        )
        slack(message: "Successfully distributed a new beta build")
    end

    desc "Deploy a new version to the Google Play Store"
    lane :release do
        supply(
            apk: "platforms/android/build/outputs/apk/android-release.apk",
            package_name: "****",
            json_key: "key.json",
            track: "production"
        )
    end

    after_all do |lane|
        unless is_ci
            # This block is called, only if the executed lane was successful
            slack()
        end
    end

    error do |lane, exception|
        if is_ci
            slack(message: exception.message, success: false)
        end
    end
end


# More information about multiple platforms in fastlane: https://github.com/fastlane/fastlane/blob/master/fastlane/docs/Platforms.md
# All available actions: https://docs.fastlane.tools/actions

# fastlane reports which actions are used
# No personal data is recorded. Learn more at https://github.com/fastlane/enhancer
```
</details>

<details><summary>`./fastlane/Appfile`</summary>

```ruby
app_identifier "********" # The bundle identifier of your app
apple_id "me@email.com" # Your Apple email address

team_id "*********"  # Developer Portal Team ID
itc_team_id "*********"
itc_team_name "***"

# FASTLANE_ITC_TEAM_ID` or `FASTLANE_ITC_TEAM_NAME

# you can even provide different app identifiers, Apple IDs and team names per lane:
# More information: https://github.com/fastlane/fastlane/blob/master/fastlane/docs/Appfile.md
```
</details>

### fastlane gems

| Gem      | Version | Update-Status |
| -------- | ------- | ------------- |
| fastlane | 2.17.1  | ✅ Up-To-Date  |

### Loaded fastlane plugins:

**No plugins Loaded**

<details><summary><b>Loaded gems</b></summary>

| Gem                       | Version      |
| ------------------------- | ------------ |
| did_you_mean              | 1.0.0        |
| executable-hooks          | 1.3.2        |
| bundler-unload            | 1.0.2        |
| rubygems-bundler          | 1.4.4        |
| bundler                   | 1.14.4       |
| io-console                | 0.4.5        |
| CFPropertyList            | 2.3.5        |
| i18n                      | 0.8.0        |
| json                      | 1.8.6        |
| minitest                  | 5.10.1       |
| thread_safe               | 0.3.5        |
| tzinfo                    | 1.2.2        |
| activesupport             | 4.2.7.1      |
| public_suffix             | 2.0.5        |
| addressable               | 2.5.0        |
| babosa                    | 1.0.2        |
| claide                    | 1.0.1        |
| fuzzy_match               | 2.0.4        |
| nap                       | 1.1.0        |
| cocoapods-core            | 1.2.0        |
| cocoapods-deintegrate     | 1.0.1        |
| cocoapods-downloader      | 1.1.3        |
| cocoapods-plugins         | 1.0.0        |
| cocoapods-search          | 1.0.0        |
| cocoapods-stats           | 1.0.0        |
| netrc                     | 0.7.8        |
| cocoapods-trunk           | 1.1.2        |
| cocoapods-try             | 1.1.0        |
| colored                   | 1.2          |
| escape                    | 0.0.4        |
| fourflusher               | 2.0.1        |
| gh_inspector              | 1.0.3        |
| molinillo                 | 0.5.6        |
| ruby-macho                | 0.2.6        |
| nanaimo                   | 0.2.3        |
| xcodeproj                 | 1.4.2        |
| cocoapods                 | 1.2.0        |
| coderay                   | 1.1.1        |
| highline                  | 1.7.8        |
| commander                 | 4.4.3        |
| unf_ext                   | 0.0.7.2      |
| unf                       | 0.1.4        |
| domain_name               | 0.5.20161129 |
| dotenv                    | 2.2.0        |
| excon                     | 0.55.0       |
| multipart-post            | 2.0.0        |
| faraday                   | 0.11.0       |
| http-cookie               | 1.0.3        |
| faraday-cookie_jar        | 0.0.6        |
| faraday_middleware        | 0.11.0.1     |
| fastimage                 | 2.0.1        |
| jwt                       | 1.5.6        |
| little-plugger            | 1.1.4        |
| multi_json                | 1.12.1       |
| logging                   | 2.1.0        |
| memoist                   | 0.15.0       |
| os                        | 0.9.6        |
| signet                    | 0.7.3        |
| googleauth                | 0.5.1        |
| httpclient                | 2.8.3        |
| hurley                    | 0.2          |
| mime-types-data           | 3.2016.0521  |
| mime-types                | 3.1          |
| uber                      | 0.0.15       |
| representable             | 2.3.0        |
| retriable                 | 2.1.0        |
| google-api-client         | 0.9.28       |
| mini_magick               | 4.5.1        |
| multi_xml                 | 0.6.0        |
| plist                     | 3.2.0        |
| rubyzip                   | 1.2.1        |
| security                  | 0.1.3        |
| slack-notifier            | 1.5.1        |
| terminal-notifier         | 1.7.1        |
| unicode-display_width     | 1.1.3        |
| terminal-table            | 1.7.3        |
| tty-screen                | 0.5.0        |
| word_wrap                 | 1.0.0        |
| rouge                     | 1.11.1       |
| xcpretty                  | 0.2.4        |
| xcpretty-travis-formatter | 0.0.4        |
| method_source             | 0.8.2        |
| slop                      | 3.6.0        |
</details>

*generated on:* **2017-02-19**
</details>

**TravisCI ENV**
```
Setting environment variables from repository settings
$ export FASTLANE_SESSION='---\n-**** \n'
$ export SPACESHIP_SESSION='---\n-**** \n'

Setting environment variables from .travis.yml
$ export DEV_PORTAL_TEAM_ID=[secure]
$ export MATCH_PASSWORD=[secure]
$ export KEYSTORE_PASSWORD=[secure]
$ export ANDROID_HOME=/usr/local/opt/android-sdk
$ export JAVA_HOME=$(/usr/libexec/java_home)
$ export MATCH_KEYCHAIN_NAME=ios-build
$ export FASTLANE_APPLE_APPLICATION_SPECIFIC_PASSWORD=[secure]
$ export FASTLANE_PASSWORD=[secure]
```

**Comment by Badmuts** [comment]
Ok I still haven't figured this one out. What I will try now is to setup a simple 'bot' account with 2-fac disabled and use that account to deploy to TestFlight or the Appstore.

**Comment by RafaelPlantard** [comment]
I'm with the same issue! 👎

**Comment by Badmuts** [comment]
Ok I got the build working by creating a new iTunes connect user with permissions to upload and deploy apps. 

Had some time to check the source code for fastlane but could not find anything related.

I think there might be an security measure in place from Apple, even though there is a session available it still asks for a auth code because of the sudden change in location. I am located in The Netherlands and Travis Mac build servers are located in Atlanta US.

I could try to start a debug build, authorise via spaceauth (`fastlane spaceauth -u user@email.com`) and use the generated session as the `FASTLANE_SESSION` env var.

At least I got it working now with 2 factor disabled but haven't figured out if this is a problem with Fastlane, Travis or Apple.

**Comment by RafaelPlantard** [comment]
@Badmuts it worked just with no TFA, right?

**Comment by Badmuts** [comment]
@RafaelPlantard Yes, by creating a new account where 2FA is not enabled I was able to upload builds to TestFlight via `pilot`. Via this method you do not need `FASTLANE_SESSION` only `FASTLANE_PASSWORD` and off course the rest of your login information via the `Appfile` or something else.

However the `FASTLANE_SESSION` workaround for 2FA is still not working for me so I am keeping this issue open.

**Comment by RafaelPlantard** [comment]
I will use your solution for while. Thanks man @Badmuts 👍

**Comment by Badmuts** [comment]
No problem @RafaelPlantard ! Hope you get things running.

For some more clarity both the methods outlined on [docs.fastlane.tools](https://docs.fastlane.tools/best-practices/continuous-integration/) and in the [spaceship readme](https://github.com/fastlane/fastlane/tree/master/spaceship#2-step-verification) did not work for me.

**Comment by RafaelPlantard** [comment]
@Badmuts for me too. Nothing works into fastlane docs (👎).

**Comment by KrauseFx** [comment]
Hey everyone, it seems like there is an issue with escaping the environment variable in Travis. We haven't had the chance to reproduce this yet to work on a fix. For now the only solution is to create a new account with limited access to use on your CI.

**Comment by Badmuts** [comment]
Thanks for the update @KrauseFx

**Comment by mfurtak** [comment]
@Badmuts Thinking about your theory around location... have you tried running the `pilot` distribution from your local machine instead of CI?

If you want to experiment with that, I would suggest ensuring that your local environment is "clean" (like you might expect on CI) by taking the steps I describe in: https://github.com/fastlane/fastlane/issues/8218#issuecomment-280862932

**Comment by mfurtak** [comment]
Hi all, I am trying to reproduce this, but so far I have not had any luck. By using a combination of `FASTLANE_PASSWORD`, `FASTLANE_APPLE_APPLICATION_SPECIFIC_PASSWORD`, and `FASTLANE_SESSION` I am able to do a _pilot_ distribution from my local machine without being prompted for any credentials.

My next step will be to try to do the same from Travis CI, and see if it is any different.

**Comment by mfurtak** [comment]
Hi @Badmuts 

It has taken a long time to get to this place (partially because of the Amazon internet disruption this week... 😅 ), but I believe I have interesting evidence from Travis CI:

```
$ export FASTLANE_SESSION=[secure]
The previous command failed, possibly due to a malformed secure environment variable.
Please be sure to escape special characters such as ' ' and '$'.
For more information, see https://docs.travis-ci.com/user/encryption-keys.
```

This comes from having set the session value through the web UI for defining environment variables. You had said you set yours through the `.travis.yml` file, correct? Do you see any output similar to the above in your runs?

**Comment by RafaelPlantard** [comment]
@mfurtak When I put into Travis CI I always put FASTLANE_SESSION between ' before save it. Understand?

**Comment by Badmuts** [comment]
Sorry for my late reaction, been very busy lately 😄 

I have encountered that issue before. I then used the travis cli with the encrypt command and made sure the special characters where escaped.

I also used an .env file to check if that would work.

I have started a debug session and printed the FASTLANE_SESSION variable. The strange thing was that I got the exact session I had locally but it was not formatted (`\n` was ignored) as it was locally. My local env uses zsh.

I was always able to run fastlane from my local machine, this might has something to do with the keychain?

I have not yet found the time to reproduce the Travis env locally which should make debugging a bit easier.

**TLDR;** I have encountered the malformed FASTLANE_SESSION error but fixed it by escaping the string as described here: https://docs.travis-ci.com/user/encryption-keys#Note-on-escaping-certain-symbols

**Comment by Badmuts** [comment]
I might think this is a problem with the travis-cli that encrypts the env var so I added an env variable without encrypting it and avoided the malformed error when I encountered these problems but still did not manage to run Fastlane without prompting for 2fa codes
